Compiler error if a global hook is within an `internal` nested class.

If you have a global hook, on an `internal` static class that is nested within another class, no errors are shown when writing the code, but trying to compile the project results in an error
> ...\obj\Debug\net9.0\TUnit.Core.SourceGenerator\TUnit.Core.SourceGenerator.CodeGenerators.TestHooksGenerator\Hooks_Inner-706b34ed66134b18b34cb2c8bdf75210.Generated.cs(30,48,30,53): error CS0122: 'Outer.Inner' is inaccessible due to its protection level

```cs
public static class Outer
{
    static class Inner
    {
        [BeforeEvery(TestDiscovery)]
        public static Task BeforeDiscovery()
        {
            return Task.CompletedTask;
        }
    }
}
```

The hook works fine if you put it in the Outer class, or make the inner class public

```cs
public static class Outer
{
    public static class Inner
    {
        [BeforeEvery(TestDiscovery)]
        public static Task BeforeDiscovery()
        {
            return Task.CompletedTask;
        }
    }

    [BeforeEvery(TestDiscovery)]
    public static Task BeforeDiscovery()
    {
        return Task.CompletedTask;
    }
}
```

--- a/TUnit.Analyzers.Tests/ClassAccessibilityAnalyzerTests.cs
+++ b/TUnit.Analyzers.Tests/ClassAccessibilityAnalyzerTests.cs
@@ -1,0 +1,56 @@
+using NUnit.Framework;
+using Verifier = TUnit.Analyzers.Tests.Verifiers.CSharpAnalyzerVerifier<TUnit.Analyzers.ClassAccessibilityAnalyzer>;
+
+namespace TUnit.Analyzers.Tests;
+
+public class ClassAccessibilityAnalyzerTests
+{
+    [Test]
+    public async Task Inner_Internal_Class_Flagged()
+    {
+        await Verifier
+			.VerifyAnalyzerAsync(
+				"""
+                using TUnit.Core;
+
+                public static class Outer
+                {
+                    static class {|#0:Inner|}
+                    {
+                        [BeforeEvery(HookType.TestDiscovery)]
+                        public static void BeforeDiscovery()
+                        {
+                        }
+                    }
+                }
+                """,
+                Verifier.Diagnostic(Rules.TypeMustBePublic)
+                    .WithLocation(0)
+			);
+    }
+    
+    [Test]
+    public async Task Public_No_Errors()
+    {
+        await Verifier.VerifyAnalyzerAsync(
+            """
+            using TUnit.Core;
+            
+            public static class Outer
+            {
+                public static class Inner
+                {
+                    [BeforeEvery(HookType.TestDiscovery)]
+                    public static void BeforeDiscovery()
+                    {
+                    }
+                }
+            
+                [BeforeEvery(HookType.TestDiscovery)]
+                public static void BeforeDiscovery()
+                {
+                }
+            }
+            """);
+    }
+}

--- a/TUnit.Analyzers/ClassAccessibilityAnalyzer.cs
+++ b/TUnit.Analyzers/ClassAccessibilityAnalyzer.cs
@@ -33,20 +33,12 @@ public class ClassAccessibilityAnalyzer : ConcurrentDiagnosticAnalyzer
         }
 
         var compilation = context.Compilation;
-
-        var testsOrHooks = namedTypeSymbol.GetMembers()
-            .OfType<IMethodSymbol>()
-            .Where(x => x.IsTestMethod(compilation) || x.IsHookMethod(compilation, out _, out _, out _))
-            .ToArray();
         
-        if (!testsOrHooks.Any())
+        if (!namedTypeSymbol.GetMembers()
+                .OfType<IMethodSymbol>()
+                .Any(x => x.IsTestMethod(compilation) || x.IsHookMethod(compilation, out _, out _, out _)))
         {
             return;
-        }
-        
-        foreach (var methodSymbol in testsOrHooks.Where(x => x.DeclaredAccessibility != Accessibility.Public))
-        {
-            context.ReportDiagnostic(Diagnostic.Create(Rules.MethodMustBePublic, methodSymbol.Locations.FirstOrDefault()));
         }
 
         if (namedTypeSymbol.ContainingType != null)

--- a/TUnit.Analyzers/ClassAccessibilityAnalyzer.cs
+++ b/TUnit.Analyzers/ClassAccessibilityAnalyzer.cs
@@ -1,0 +1,57 @@
+﻿using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using TUnit.Analyzers.Extensions;
+
+namespace TUnit.Analyzers;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public class ClassAccessibilityAnalyzer : ConcurrentDiagnosticAnalyzer
+{
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } =
+        ImmutableArray.Create
+        (
+            Rules.MethodMustBePublic,
+            Rules.TypeMustBePublic
+        );
+
+    protected override void InitializeInternal(AnalysisContext context)
+    { 
+        context.RegisterSymbolAction(AnalyzeSymbol, SymbolKind.NamedType);
+    }
+    
+    private void AnalyzeSymbol(SymbolAnalysisContext context)
+    { 
+        if (context.Symbol is not INamedTypeSymbol namedTypeSymbol)
+        {
+            return;
+        }
+
+        if (namedTypeSymbol.DeclaredAccessibility == Accessibility.Public)
+        {
+            return;
+        }
+
+        var compilation = context.Compilation;
+
+        var testsOrHooks = namedTypeSymbol.GetMembers()
+            .OfType<IMethodSymbol>()
+            .Where(x => x.IsTestMethod(compilation) || x.IsHookMethod(compilation, out _, out _, out _))
+            .ToArray();
+        
+        if (!testsOrHooks.Any())
+        {
+            return;
+        }
+        
+        foreach (var methodSymbol in testsOrHooks.Where(x => x.DeclaredAccessibility != Accessibility.Public))
+        {
+            context.ReportDiagnostic(Diagnostic.Create(Rules.MethodMustBePublic, methodSymbol.Locations.FirstOrDefault()));
+        }
+
+        if (namedTypeSymbol.ContainingType != null)
+        {
+            context.ReportDiagnostic(Diagnostic.Create(Rules.TypeMustBePublic, namedTypeSymbol.Locations.FirstOrDefault()));
+        }
+    }
+}

--- a/TUnit.Analyzers/Resources.Designer.cs
+++ b/TUnit.Analyzers/Resources.Designer.cs
@@ -1408,5 +1408,32 @@ namespace TUnit.Analyzers {
                 return ResourceManager.GetString("TUnit0050Title", resourceCulture);
             }
         }
+        
+        /// <summary>
+        ///   Looks up a localized string similar to Type must be public..
+        /// </summary>
+        internal static string TUnit0051Description {
+            get {
+                return ResourceManager.GetString("TUnit0051Description", resourceCulture);
+            }
+        }
+        
+        /// <summary>
+        ///   Looks up a localized string similar to Type must be public.
+        /// </summary>
+        internal static string TUnit0051MessageFormat {
+            get {
+                return ResourceManager.GetString("TUnit0051MessageFormat", resourceCulture);
+            }
+        }
+        
+        /// <summary>
+        ///   Looks up a localized string similar to Type must be public.
+        /// </summary>
+        internal static string TUnit0051Title {
+            get {
+                return ResourceManager.GetString("TUnit0051Title", resourceCulture);
+            }
+        }
     }
 }

--- a/TUnit.Analyzers/Resources.resx
+++ b/TUnit.Analyzers/Resources.resx
@@ -468,4 +468,13 @@
     <data name="TUnit0050Title" xml:space="preserve">
         <value>Too many arguments provided</value>
     </data>
+    <data name="TUnit0051Description" xml:space="preserve">
+        <value>Type must be public.</value>
+    </data>
+    <data name="TUnit0051MessageFormat" xml:space="preserve">
+        <value>Type must be public</value>
+    </data>
+    <data name="TUnit0051Title" xml:space="preserve">
+        <value>Type must be public</value>
+    </data>
 </root>

--- a/TUnit.Analyzers/Rules.cs
+++ b/TUnit.Analyzers/Rules.cs
@@ -120,6 +120,9 @@ public static class Rules
     public static readonly DiagnosticDescriptor TooManyArguments =
         CreateDescriptor("TUnit0050", UsageCategory, DiagnosticSeverity.Error);
     
+    public static readonly DiagnosticDescriptor TypeMustBePublic =
+        CreateDescriptor("TUnit0051", UsageCategory, DiagnosticSeverity.Error);
+    
     private static DiagnosticDescriptor CreateDescriptor(string diagnosticId, string category, DiagnosticSeverity severity)
     {
         return new DiagnosticDescriptor(

